### PR TITLE
[Bugfix] Fix the uninitialized pin_memory flag for empty graph

### PR DESCRIPTION
**Pull request by chang-l**
## Description
This is to fix https://github.com/dmlc/dgl/issues/5592. For future/own reference, I will detail the root cause in the following.

The root cause is when the first relational graph is empty, the entire heterograph could return `is_pinned() = true` even if they are never being pinned by DGL or PyTorch at all. 

This is because (1) the empty unit graph would return `is_pinned = True` due to the following init. logic: https://github.com/dmlc/dgl/blob/ae8cbde5cfa699ff8d9e774ae6acd2cb8f7b451a/include/dgl/aten/coo.h#L67-L69 and (2) the first relational graph determines if the entire hetero-graph is pinned. 
https://github.com/dmlc/dgl/blob/ae8cbde5cfa699ff8d9e774ae6acd2cb8f7b451a/src/graph/heterograph.h#L58

As a result, when the sub-graph is pickled/unpickled by workers/main process after sampling, the code will always try to pin the (sub)graph as long as its `relation_graphs_[0]` is an empty graph (`coo.data`, `coo.row` and `coo.col` are NullArray). See here https://github.com/dmlc/dgl/blob/ae8cbde5cfa699ff8d9e774ae6acd2cb8f7b451a/src/graph/pickle.cc#L64 is for sub-graph pickling, which will always write `True` if `graph->relation_graphs_[0]` is empty. And then here https://github.com/dmlc/dgl/blob/ae8cbde5cfa699ff8d9e774ae6acd2cb8f7b451a/src/graph/pickle.cc#L156-L158 will pin the unpickled graph even if it never gets pinned before.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [x] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [x] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [x] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

## Changes
- If its a null/empty graph, always init. the `is_pinned` to false
- Update tests to cover this case

### CODE REVIEW

**Comment by chang-l** [comment]
@Rhett-Ying  Can you please try this PR to see if it can fix your crash? Also, what PyTorch version did you use for testing?

**Comment by Rhett-Ying** [comment]
@dgl-bot run r6i.16xlarge bench_rgcn_hetero dmlc/1.0.x

**Comment by Rhett-Ying** [comment]
@chang-l   pytorch: 1.13.1

**Comment by Rhett-Ying** [comment]
@chang-l confirmed. it works now.

File: include/dgl/aten/coo.h [hunk 1]
```diff
@@ -64,9 +64,12 @@ struct COOMatrix {
         data(darr),
         row_sorted(rsorted),
         col_sorted(csorted) {
-    is_pinned = (aten::IsNullArray(row) || row.IsPinned()) &&
-                (aten::IsNullArray(col) || col.IsPinned()) &&
-                (aten::IsNullArray(data) || data.IsPinned());
+    if (!aten::IsNullArray(row) || !aten::IsNullArray(col) ||
```
Comment (Rhett-Ying) [inline include/dgl/aten/coo.h:67]: how about wrapping these `IsNullArray()` check into a member function of `COOMatrix`: `IsEmpty()`to remove code redundancy?

**Comment by chang-l** [comment]
> pytorch: 1.13.1

Thanks. In fact, I was not able to reproduce the crash with PyTorch > 1.12 (built/run with cuda) yesterday... I guess you ran it with CPU-only? (of course it is a bug even if it ran through with cuda)
Anyway, glad it fixes the issue. I will finalize this PR later today :)

**Comment by Rhett-Ying** [comment]
@chang-l yes, I hit this issue when run on CPU.

**Comment by Rhett-Ying** [comment]
@dgl-bot run r6i.16xlarge bench_rgcn_hetero dmlc/1.0.x

File: include/dgl/aten/coo.h [hunk 2]
```diff
@@ -138,16 +145,21 @@ struct COOMatrix {
...
+          aten::IsNullArray(data) ? data : data.PinMemory(), row_sorted,
+          col_sorted);
+      CHECK(new_coo.is_pinned)
+          << "An internal DGL error has occured while trying to pin a COO "
+             "matrix. Please file a bug at "
+             "'https://github.com/dmlc/dgl/issues' "
+             "with the above stacktrace.";
+      return new_coo;
+    }
+    is_pinned = true;
```
Comment (yaox12) [inline include/dgl/aten/coo.h:161]: `is_pinned` will be set to true even if `IsEmpty() == true`, is it expected?

Comment (chang-l) [reply include/dgl/aten/coo.h:161]: You're right. `is_pinned` should not be set for this pinning method :( 
Thank you!

File: include/dgl/aten/coo.h [hunk 3]
```diff
@@ -159,13 +170,17 @@ struct COOMatrix {
...
+    if (!IsEmpty()) {
+      if (is_pinned) return;
+      row.PinMemory_();
+      col.PinMemory_();
+      if (!aten::IsNullArray(data)) {
+        data.PinMemory_();
+      }
+      is_pinned = true;
     }
     is_pinned = true;
```
Comment (yaox12) [inline include/dgl/aten/coo.h:182]: Also here.

File: include/dgl/aten/csr.h [hunk 4]
```diff
@@ -151,13 +162,17 @@ struct CSRMatrix {
...
+    if (!IsEmpty()) {
+      if (is_pinned) return;
+      indptr.PinMemory_();
+      indices.PinMemory_();
+      if (!aten::IsNullArray(data)) {
+        data.PinMemory_();
+      }
+      is_pinned = true;
     }
     is_pinned = true;
```
Comment (yaox12) [inline include/dgl/aten/csr.h:174]: And here.

Comment (chang-l) [reply include/dgl/aten/csr.h:174]: This is probably necessary I think, as it is an in-place method. We should flag it even if it is empty when we called this function.

File: include/dgl/aten/coo.h (hunk 2 above)
Comment (chang-l) [reply include/dgl/aten/coo.h]: Given the test failures, now, I start thinking I probably should set `is_pinned ` to True even if `IsEmpty() == true`, for both in-place and out-of-place methods. If graph is empty, we should just change the flag in these functions. Then, out-of-place method becomes the same as in-place method as there are no memory movements, but just flag changing.

**Review by Rhett-Ying — APPROVED** [review]

**Comment by Rhett-Ying** [comment]
@chang-l  are we ready to merge this PR?

**Pull request merged**